src/gadgets/encode_image.rs

```diff
@@ -0,0 +1,134 @@
+}
+
+impl<F: PrimeFieldBits> EncodeImageChip<F> {
+    pub fn construct(config: EncodeImageChipConfig, binarization_thresholds: Array3<u16>) -> Self {
+        let greater_than_chip = GreaterThanChip::construct(config.greater_than_chip_config.clone());
+        Self {
+            greater_than_chip,
+            config,
+            binarization_thresholds,
+        }
+    }
+
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        x: Column<Advice>,
+        y: Column<Advice>,
+        diff: Column<Advice>,
+        is_gt: Column<Advice>,
+        byte_column: TableColumn,
+    ) -> EncodeImageChipConfig {
+        let greater_than_chip_config =
+            GreaterThanChip::configure(meta, x, y, diff, is_gt, byte_column);
+        EncodeImageChipConfig {
+            advice_column: is_gt,
+            greater_than_chip_config,
+        }
+    }
+}
+
+impl<F: PrimeFieldBits> EncodeImageInstructions<F> for EncodeImageChip<F> {
+    fn encode_image(
+        &self,
+        mut layouter: impl Layouter<F>,
+        image: &Array2<u8>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        let (width, height) = (image.shape()[0], image.shape()[1]);
+
+        let mut intensity_cells: BTreeMap<(usize, usize), AssignedCell<F, F>> = BTreeMap::new();
+        let mut bit_cells = vec![];
+
+        for b in 0..self.binarization_thresholds.shape()[2] {
+            for i in 0..width {
+                for j in 0..height {
+                    let threshold = self.binarization_thresholds[(i, j, b)];
+                    assert!(threshold <= 256);
+
+                    let bit_cell = if threshold == 0 {
+                        // If the threshold is zero, the bit is always one, regardless of the of the intensity.
+                        // Unfortunately, this has to be handled separately, as the greater than gadget can't
+                        // handle a threshold of -1 (see below).
+                        layouter.assign_region(
+                            || "bit is one",
+                            |mut region| {
+                                region.assign_advice_from_constant(
+                                    || "gt",
+                                    self.config.advice_column,
+                                    0,
+                                    F::ONE,
+                                )
+                            },
+                        )?
+                    } else {
+                        // The result should be true if the intensity is greater or equal than the threshold,
+                        // but the gadget only implements greater than, so we need to subtract 1 from the threshold.
+                        // Because we already handled the threshold == 0 case, this means that `t` is now in the
+                        // range [0, 255], which is required by the greater than gadget.
+                        let t = F::from((self.binarization_thresholds[(i, j, b)] - 1) as u64);
```

Comment:
You've resolved the conversation, but just to be sure, because we're using the thresholds with `assign_advice_from_constant()`, it will be copied to a public fixed column and committed to during key generation (and equality constraints ensure the same values are used in the advice cells).

In other words, making sure the correct thresholds are used is part of validating the verification key, which is computed from the (public) model.